Make `TargetPlatform` a managed type

fourlastor opened the pull request:
This PR converts `TargetPlatform` to use [managed properties](https://docs.gradle.org/current/userguide/custom_gradle_types.html#managed_properties).

In short, this means that `TargetPlatform` is not instantiated directly in the codebase, but by Gradle (which fill in the various properties for us)

In addition, I changed the holding type for `RuntimePluginExtension.targetPlatforms` to [NamedDomainObjectContainer](https://docs.gradle.org/7.6/javadoc/org/gradle/api/NamedDomainObjectContainer.html) which is basically the default "map like" container for gradle plugins.

fourlastor commented on src/main/groovy/org/beryx/runtime/data/TargetPlatform.groovy:
```diff
@@ -16,60 +16,42 @@
     }
 
-    LazyString jdkDownload(String downloadUrl, Closure downloadConfig = null) {
+    Provider<String> jdkDownload(String downloadUrl, Closure downloadConfig = null) {
```
This is another changed I made, since this was an internal type. It changed the approach to rely on `Provider<String>` but in practice it should be equivalent

hakanai commented on src/main/groovy/org/beryx/runtime/data/RuntimePluginExtension.groovy:
```diff
@@ -39,7 +40,7 @@ class RuntimePluginExtension {
     final ListProperty<String> modules
     final Property<String> javaHome
-    final Provider<Map<String, TargetPlatform>> targetPlatforms
+    final NamedDomainObjectContainer targetPlatforms
```
Missing `<TargetPlatform>` here possibly?

fourlastor replied:
[e61b4c3](https://github.com/beryx/badass-runtime-plugin/pull/140/commits/e61b4c39cbdf8bc323a7155e8499ee7754165cd8) yep, added!

hakanai commented on src/main/groovy/org/beryx/runtime/data/TargetPlatform.groovy:
```diff
@@ -16,60 +16,42 @@
+import org.gradle.api.provider.ListProperty
+import org.gradle.api.provider.Property
+import org.gradle.api.provider.Provider
 
```
Was there a particular reason for removing `@CompileStatic`?

fourlastor replied:
[e61b4c3](https://github.com/beryx/badass-runtime-plugin/pull/140/commits/e61b4c39cbdf8bc323a7155e8499ee7754165cd8) nope, this was me misunderstanding what it was for (I assumed it was similar to `JvmStatic` in Kotlin), added back

hakanai reviewed (changes requested):
Seems straightforward enough, only minor comments from me.

hakanai reviewed (approved):
With changes looks good to me.

Let us know when you're done testing it in real usage, then this one should be good to go.

fourlastor commented:
After testing it, I had to change a few things:

1. `abstract class` is not good enough to be able to define abstract properties, it would need to be an interface (it fails when trying to run the task). I reverted the changes here and made it back as a concrete class which is instantiated
2. `TargetPlatform` is not `Serializable` anymore (as `Property` appears to be not serializable), I removed the interface implementation and used the `@Nested` annotation for the tasks input, which in short looks at properties annotated with `@Input` and similar to determine the task inputs/outputs

hakanai commented:
If you're fine with the current state, I think we're good to go?

fourlastor commented:
Yep, good to go for me 👍🏻
